// @issr/core/lib/cjs/effect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Effect = exports.Statuses = void 0;
var Statuses;
(function (Statuses) {
    Statuses["wait"] = "wait";
    Statuses["done"] = "done";
    Statuses["failed"] = "failed";
})(Statuses = exports.Statuses || (exports.Statuses = {}));
var Effect = /** @class */ (function () {
    function Effect(_a) {
        var _this = this;
        var id = _a.id;
        this.getStatus = function () { return _this.status; };
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        this.getCallback = function () { return _this.callback; };
        this.done = function () {
            _this.status = Statuses.done;
        };
        this.failed = function () {
            _this.status = Statuses.failed;
        };
        this.getId = function () { return _this.id; };
        this.id = id;
        this.status = Statuses.wait;
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    Effect.prototype.addCallback = function (cb) {
        if (this.status === Statuses.wait) {
            this.callback = cb;
        }
    };
    return Effect;
}());
exports.Effect = Effect;
